File: src/lib/moodMatrix.ts
/**
 * #472: Pure helpers for the before/after mood matrix.
 * Extracted from MoodMatrix.tsx so vitest (node mode, jsx:preserve) can import
 * and test them without a JSX transform.
 */

export const MOOD_KEYS = ["calm", "focus", "energy", "anxiety", "overall"] as const;
export type MoodKey = (typeof MOOD_KEYS)[number];

export const MOOD_LABELS: Record<MoodKey, string> = {
  calm: "Calm",
  focus: "Focus",
  energy: "Energy",
  anxiety: "Anxiety",
  overall: "Overall",
};

export type MoodEntry = { before: number | null; after: number | null };
export type MoodMatrixValue = Partial<Record<MoodKey, MoodEntry>>;

/** Returns true if the matrix contains at least one non-null cell. */
export function isMoodMatrixTouched(value: MoodMatrixValue): boolean {
  return MOOD_KEYS.some((k) => {
    const e = value[k];
    return e !== undefined && (e.before !== null || e.after !== null);
  });
}

/** Strips un-touched rows (both before and after are null) to keep the
 *  payload lean — rows where the user never tapped anything are omitted. */
export function buildMoodMatrixPayload(
  value: MoodMatrixValue,
): Record<string, { before: number | null; after: number | null }> {
  const out: Record<string, { before: number | null; after: number | null }> = {};
  for (const key of MOOD_KEYS) {
    const e = value[key];
    if (e && (e.before !== null || e.after !== null)) {
      out[key] = e;
    }
  }
  return out;
}

File: ios/StillPointShared/Sources/StillPointShared/MoodMatrix.swift
import Foundation

/// #472 / #635: before/after mood matrix keys and helpers.
/// Mirrors `src/lib/moodMatrix.ts` for web-iOS parity.
public enum MoodKey: String, CaseIterable, Codable, Sendable {
    case calm
    case focus
    case energy
    case anxiety
    case overall

    public var label: String {
        switch self {
        case .calm: return "Calm"
        case .focus: return "Focus"
        case .energy: return "Energy"
        case .anxiety: return "Anxiety"
        case .overall: return "Overall"
        }
    }
}

public struct MoodMatrixEntry: Codable, Sendable, Equatable {
    public let before: Int?
    public let after: Int?

    public init(before: Int? = nil, after: Int? = nil) {
        self.before = before
        self.after = after
    }

    /// Server validation requires both `before` and `after` keys (null allowed).
    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(before, forKey: .before)
        try container.encode(after, forKey: .after)
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        before = try container.decodeIfPresent(Int.self, forKey: .before)
        after = try container.decodeIfPresent(Int.self, forKey: .after)
    }

    private enum CodingKeys: String, CodingKey {
        case before, after
    }
}

public enum MoodMatrixLogic {
    /// Returns true when the matrix contains at least one non-nil cell.
    public static func isTouched(_ value: [MoodKey: MoodMatrixEntry]) -> Bool {
        MoodKey.allCases.contains { key in
            guard let entry = value[key] else { return false }
            return entry.before != nil || entry.after != nil
        }
    }

    /// Strips rows where both before and after are nil — only touched rows are sent.
    public static func buildPayload(from value: [MoodKey: MoodMatrixEntry]) -> [String: MoodMatrixEntry] {
        var out: [String: MoodMatrixEntry] = [:]
        for key in MoodKey.allCases {
            guard let entry = value[key], entry.before != nil || entry.after != nil else { continue }
            let validated = validatedEntry(before: entry.before, after: entry.after)
            guard validated.before != nil || validated.after != nil else { continue }
            out[key.rawValue] = validated
        }
        return out
    }

    public static func isKnownKey(_ key: String) -> Bool {
        MoodKey(rawValue: key) != nil
    }

    /// Clamps integer mood cells to 1…5; out-of-range values become nil.
    public static func clampMoodValue(_ value: Int?) -> Int? {
        guard let value else { return nil }
        guard value >= 1, value <= 5 else { return nil }
        return value
    }

    public static func validatedEntry(before: Int?, after: Int?) -> MoodMatrixEntry {
        MoodMatrixEntry(
            before: clampMoodValue(before),
            after: clampMoodValue(after)
        )
    }

    /// Filters unknown keys and clamps cell values when decoding persisted data.
    public static func sanitizedStored(_ stored: [String: MoodMatrixEntry]) -> [MoodKey: MoodMatrixEntry] {
        var out: [MoodKey: MoodMatrixEntry] = [:]
        for (rawKey, entry) in stored {
            guard let key = MoodKey(rawValue: rawKey) else { continue }
            let validated = validatedEntry(before: entry.before, after: entry.after)
            if validated.before != nil || validated.after != nil {
                out[key] = validated
            }
        }
        return out
    }
}

/// #635: nil-omitting PATCH body `{ "moodMatrix": { … } }` for session recap saves.
public struct MoodMatrixPatch: Encodable, Sendable {
    public let entries: [String: MoodMatrixEntry]

    public init(entries: [String: MoodMatrixEntry]) {
        self.entries = entries
    }

    public init(from value: [MoodKey: MoodMatrixEntry]) {
        self.entries = MoodMatrixLogic.buildPayload(from: value)
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(entries, forKey: .moodMatrix)
    }

    private enum CodingKeys: String, CodingKey {
        case moodMatrix
    }
}
